### DISCUSSION TITLE
Graphviz Feature "HTML Like Tables" on Ubuntu 22.04 and Android

### DISCUSSION THREAD

**Discussion by ThomasJez** [hpcc-systems/hpcc-js-wasm#151] [General]
The Graphviz feature "HTML Like Tables" doesn't work correctly with the preinstalled version of Firefox (the snap version) of the actual LTS (22.04) of Ubuntu. For boxes with text, the text is too large and partly misplaced outside the boxes (scrrenshot).

To reproduce, take a fresh install of Ubuntu 22.04 (or boot it from an installation media), go to https://observablehq.com/@gordonsmith/graphviz and scroll down to "HTML Like Tables".

The problem doesn't occur with the tarball version, the repo-version (ppa:mozillateam/ppa) or the flatpack version of Firefox and I coudn't find any other browser on Windows or Ubuntu where this problem occurs.

But I also tested it on two Android phones (Nexus 5X Android 8.1 with Firefox and Chrome and Mi A3 with Opera and Chrome) and on Pixel 5 API 30 emulator of Android Studio with Webview Browser Tester, there this problem also occurs (with all tested browsers).

![Auswahl_001](https://user-images.githubusercontent.com/1399641/180254086-f2d5ed2e-4679-467c-815e-ba073815ecca.png)

**Comment by peterdd** [comment]
Sounds like the table cell width calculation uses another font/font-size then what the browser uses for displaying text.

**Comment by GordonSmith** [comment]
@ThomasJez could you grab the generated SVG from the broken Firefox and a working one and compare?  

My assumption is that the SVG will be identical, which probably means that the issue may be with the browsers default "Times,Serif" font used when rendering the SVG?

**Comment by GordonSmith** [comment]
Also how does it look if you change the Default Font to Arial?
```
digraph structs {
    node [shape=plaintext fontname="Arial"]
    struct1 [label=<
...
```

**Comment by ThomasJez** [comment]
The SVG are identical and adding 'fontname="Arial"' fixes the error. (I have tested it in Chrome on Android too and there it works too.)

**Comment by ThomasJez** [comment]
Additionally I have generated a SVG from the dot from https://observablehq.com/@gordonsmith/graphviz (without the added font specification) with the dot app from the graphviz. 

This uses the Times, serif font but works nonetheless in all browsers. The point positions and other numbers are diffent from the hpcc generated SVG.

Graphviz-original generated:
![graphviz-ori](https://user-images.githubusercontent.com/1399641/180441012-15d57576-ac4f-471f-a0be-4e71cde2a40c.svg)
hpcc generated:
![hpcc](https://user-images.githubusercontent.com/1399641/180441023-2a9aeedf-53fb-4a60-82c3-8c8c6ed7d399.svg)

**Comment by ThomasJez** [comment]
But anyway, specifying the font solves the problem for me and making this hint part of the documentation probably solves it for everybody.

**Comment by GordonSmith** [comment]
Finally if you press F12 and select some overflowing text in a "broken" browser, what font is it using on the CSS Computed page?
![image](https://user-images.githubusercontent.com/966863/180441729-066b286c-5cd1-4fa4-8532-f606b66c8e44.png)

Note:  The command line dot program has the ability to check local folders for fonts and use the actual font info to calc the sizes, the wasm version can only do this for a handful of fonts that are included at build time...

**Comment by ThomasJez** [comment]
As there are a lot of font parameters I am just posting the screenshot:

![VirtualBox_ClearJellyMate_22_07_2022_15_09_14](https://user-images.githubusercontent.com/1399641/180446684-4cbbc031-2cad-42f3-afba-c993de7b1da5.png)

**Comment by borango** [comment]
For your information: I have seen similar problems with too-large text an different browsers (or other user agents that can display SVG), just with using the dot binary (not even the Javascript ports). My findings were always that user agents take a lot of liberty in interpreting fonts. The root problem is: **setting a font by name, style and size is not deterministic**. There are variations at runtime. The case here does not seem to be caused by hpcc-wasm.
